### PR TITLE
Refactor io module for proper cross-platform support

### PR DESCRIPTION
**Pull request by olafurpg**
The io module in 1.8 has a lot of problems. It does now work properly on Windows, AbsolutePath has very limited methods (e.g. no .resolveSibling, .getParent, .endsWith, .startsWith), it's very inefficient (no prefix sharing for AbsolutePaths in same directory). This PR refactors AbsolutePath and RelativePath to become small wrappers around java.nio.file.Path, addressing all of the limitations of the old Absolute/RelativePath. This PR also implements nio.Path and io.File in Scala.js using Node.js APIs.

### CODE REVIEW

**Comment by olafurpg** [comment]
The tests seems to pass now on windows 👏 This PR still needs more work to clean up before code review, but in general I'm very happy with the changes and it addresses most of the issues I had with the current io module, for example #885

**Comment by olafurpg** [comment]
This PR is ready for review @xeno-by

**Comment by olafurpg** [comment]
Nice to see that the test from ae4cbd5 that was intended to fail on Windows did indeed fail on Windows
```scala
[info] - s.Attributes.filename has no Windows slash (\) *** FAILED ***
[info]   "scalahost\integration\src\main\scala\example\Example.scala" contained '\' (CrossPlatformSemanticSuite.scala:38)
```
I don't have a local windows computer to reproduce. b2be956 should fix the issue.

File: scalameta/io/js/src/main/scala/java/io/File.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.net.URI
+import java.nio.file.Path
+import scala.meta.internal.io.FileIO
+import scala.meta.internal.io.JSFs
+import scala.meta.internal.io.JSPath
+import scala.meta.internal.io.JSStats
+import scala.meta.internal.io.NodeNIOPath
+import scala.meta.internal.io.PathIO
+import scala.meta.internal.io.PlatformFileIO
```
Comment (xeno-by) [inline scalameta/io/js/src/main/scala/java/io/File.scala]: Why not simply say `import scala.meta.internal.io._`

File: scalameta/io/js/src/main/scala/java/io/File.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+// obtained implementation by experimentation on the JDK.
```
Comment (xeno-by) [inline scalameta/io/js/src/main/scala/java/io/File.scala]: @sjrd @densh What are the legal implications of doing something like this?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala [hunk 3]
```diff
@@ -42,14 +42,15 @@ package object semantic {
...
-          val (spath, scontents) = minput match {
+          val (splatformpath, scontents) = minput match {
             case mInput.File(path, charset) if charset == Charset.forName("UTF-8") =>
               path.toRelative(sourceroot).toString -> ""
             case mInput.LabeledString(label, contents) =>
               label -> contents
             case other =>
               sys.error(s"bad database: unsupported input $other")
           }
+          val spath = PathIO.toUnix(splatformpath)
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: What will this return on Windows?

File: scalameta/io/js/src/main/scala/java/io/File.scala (hunk 2 above)
Comment (sjrd) [inline scalameta/io/js/src/main/scala/java/io/File.scala]: That's perfectly fine. It means someone reverse-engineered the spec by observing the behavior, and then reimplemented that spec. That's the very best incarnation of a clean room implementation.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala (hunk 3 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: A unix filename would have forward slashes /

Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: How would `c:\foo.scala` be translated? Can we really say that `c:/foo.scala` is a unix path?

Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: The path is relative at this point so it should not be a problem.

**Comment by xeno-by** [comment]
LGTM

**Pull request merged**